astropy/table/tests/test_table.py

```diff
@@ -3023,3 +3038,14 @@ def test_remove_columns_invalid_names_messages():
     with pytest.raises(KeyError,
                        match='columns {\'[de]\', \'[de]\'} do not exist'):
         t.remove_columns(['c', 'd', 'e'])
+
+
+def test_read_write_tilde_path(home_is_tmpdir):
+    test_file = os.path.join('~', 'test.csv')
+    t1 = Table()
+    t1.add_column(Column(name='a', data=[1, 2, 3]))
+    t1.write(test_file)
+    t2 = Table.read(test_file)
+    assert np.all(t2['a'] == [1, 2, 3])
+    # Ensure the data wasn't written to the literal tilde-prefixed path
+    assert not os.path.exists(test_file)
```

Comment:
This one blows my mind a little. So, `os.path.exists` is immune to `monkeypatch` fixture thingy?

Reply:
The entire `os` module doesn't ever expand `~` by default (afaik), nor does the builtin `open()`, which is why `expanduser` has to be called explicitly by us

Reply:
So that is to say, `os.path.exists` sees `~/file.csv` as a relative path containing a directory whose name is a literal `~`, and the monkey-patched environment variables never factor in